src/field/FieldPropagator.i.hh

```diff
@@ -14,194 +16,135 @@ namespace celeritas
  */
 template<class DriverT>
 CELER_FUNCTION
-FieldPropagator<DriverT>::FieldPropagator(GeoTrackView*            track,
-                                          const ParticleTrackView& particle,
-                                          DriverT&                 driver)
-    : track_(track), driver_(driver)
+FieldPropagator<DriverT>::FieldPropagator(const ParticleTrackView& particle,
+                                          GeoTrackView*            track,
+                                          DriverT*                 driver)
+    : track_(*track), driver_(*driver)
 {
+    CELER_ASSERT(track && driver);
     CELER_ASSERT(particle.charge() != zero_quantity());
 
     using MomentumUnits = OdeState::MomentumUnits;
 
     state_.pos = track->pos();
+    state_.mom = detail::ax(value_as<MomentumUnits>(particle.momentum()),
+                            track->dir());
+}
 
-    for (size_type i = 0; i != 3; ++i)
-    {
-        state_.mom[i] = value_as<MomentumUnits>(particle.momentum())
-                        * track->dir()[i];
-    }
+//---------------------------------------------------------------------------//
+/*!
+ * Propagate a charged particle until it hits a boundary.
+ */
+template<class DriverT>
+CELER_FUNCTION auto FieldPropagator<DriverT>::operator()() -> result_type
+{
+    return (*this)(numeric_limits<real_type>::infinity());
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Propagate a charged particle in a magnetic field and update the field
- * state.
+ * Propagate a charged particle in a magnetic field.
  *
  * It utilises a magnetic field driver based on an adaptive step
  * control to track a charged particle until it travels along a curved
  * trajectory for a given step length within a required accuracy or intersects
  * with a new volume (geometry limited step).
+ *
+ * The position of the internal OdeState `state_` should be consistent with the
+ * geometry `track_`'s position, but the geometry's direction will be a series
+ * of "trial" directions that are the chords between the start and end points
+ * of a curved substep through the field. At the end of the propagation step,
+ * the geometry state's direction is updated based on the actual value of the
+ * calculated momentum.
+ *
+ * Caveats:
+ * - Due to boundary fuzziness, the track may translate slightly when it
+ *   intersects a boundary.
+ * - Due to minimum driver step length, the resulting distance moved may not
+ *   exactly add up to the updated track position (TODO: should we do a
+ *   "linear" propagation step when this happens?)
```

Comment:
I think so.  When the last trial of sub-step is within the required tolerance and the final position is not crossed a target boundary), the remaining distance difference should be pushed to the next boundary along the final direction so that the end position should be exactly on the boundary - I thought that we are doing this, i.e., using the intersect position instead of the final ending position from the driver.

Reply:
OK, that sounds right to me. I was confused by the original behavior of the Field propagator, which (I think) was the following:
- A user-requested step less than the driver minimum resulted in an infinite-step result and no boundary crossing.
- If the substep's chord has a boundary crossing, the trial endpoint steps are iterated until an intercept is found within tolerance *or* an abort counter reaches zero.
- If the driver-calculated substeps accumulate to `taken + min_step > requested` then we don't step all the way to the user-requested value.

I don't think the first bullet point is the correct behavior (which is what the TODO is about). So I could remove the `result.distance += remaining;` line, but that would result in a zero-distance step if the requested distance is too small. I don't think that's right. If we're near a boundary and the `remaining` distance gets too small, it may mean that our initial chord "cut a corner" and we don't end up intersecting a boundary along the curved path.

Ugh, I also just realized that by adding the `near_boundary` condition, the loop no longer guarantees that `remaining` is monotonically decreasing, so there's the possibility of an infinite loop again. We should add a legitimate "corner" case to the test to make sure everything is good.

Reply:
OK, I've refactored the algorithm again so it's guaranteed convergent. There will be some "fuzziness" near the boundary, but with this algorithm everything is convergent and consistent.

Compared to the original, the number of calls to `driver.advance` is usually the same but occasionally higher, although the total stepping distance is smaller. Here's one example near a boundary crossing, where the original makes the following attempts and finally steps:
* Attempt step 0.239298
* Attempt step 0.213842 => 0.213842
* Attempt step 0.215639 => 0.215639
* Substep 0.215511 => 0.215511

in contrast, the new algorithm does additional "safe" substeps that reduce the total remaining step length, allowing for smaller trial steps:
* Attempt step 0.239298
* Substep 0.213842
* Attempt step 0.025456
* Substep 0.00166614
* Attempt step 0.0237899
* Substep 1.20314e-05
* Attempt step 0.0237778